Group pages update and re-skin

dmitryduev opened the pull request:
In this PR:

- Links to Sources and Groups in Sidebar
- Re-skin Group, GroupList and Groups components with more MUI
- Alert streams/filters handling on Group page
- Redirect to groups list page after deleting group on single group view page
- Fix relevant frontend tests, add test of adding/deleting stream/filter to group
- Back-end adjustments to ensure user permissions compliance

![groups-01](https://user-images.githubusercontent.com/7557205/89997950-b839c480-dc41-11ea-9633-936b6cc4d387.gif)

dmitryduev commented on skyportal/tests/conftest.py:
```diff
@@ -8,15 +8,13 @@
+from baselayer.app.test_util import driver  # noqa: F401
+from baselayer.app.test_util import MyCustomWebDriver  # noqa: F401
+from baselayer.app.test_util import set_server_url
+from baselayer.app.test_util import reset_state  # noqa: F401
```
Had to do it like this, as `flake8` would otherwise ignore the `noqa`'s (tried putting them in everywhere..)

dmitryduev replied:
Yes, errors get thrown if these are removed.

acrellin commented on static/js/components/Group.jsx:
```diff
@@ -47,61 +200,378 @@ const Group = () => {
+                            </ListItem>
+                          ) : (
+                            ""
+                          ) /* <div style={{paddingLeft: "2em"}}>No filters defined on stream</div> */
```
Should this commented-out line still be here?

dmitryduev replied:
Removed

acrellin commented on static/js/components/Group.jsx:
```diff
@@ -47,61 +200,378 @@ const Group = () => {
     );
   }
-  return <div>Loading group</div>;
+  return (
```
Slight preference for placing the `return <LoadingPlaceholderThingy />` statement above the main return statement (and inverting the check accordingly if relevant) but this is fine. I like this spinner -- we should adopt this in more places throughout the app!

dmitryduev replied:
Done

acrellin commented on static/js/ducks/filter.js:
```diff
@@ -0,0 +1,104 @@
+export const DELETE_GROUP_FILTER = "skyportal/DELETE_GROUP_FILTER";
+export const DELETE_GROUP_FILTER_OK = "skyportal/DELETE_GROUP_FILTER_OK";
+
+export const FETCH_FILTER_V = "skyportal/FETCH_FILTER_V";
```
Does V stand for version here? Perhaps make that more explicit for folks like me who don't know what that means

dmitryduev replied:
Forgot to remove this from this duck -- this is relevant for F, so will be part of a PR on F, with proper naming.

acrellin commented on static/js/ducks/filter.js:
```diff
@@ -0,0 +1,104 @@
+const reducer = (state = {}, action) => {
+  switch (action.type) {
+    case FETCH_FILTER_OK: {
+      return action.data;
```
What are the keys of `action.data`? It would be helpful to make it more explicit here, as is done in other reducers (e.g. grab the keys/vals in a destructuring assignment and then `return { key1, key2 };` or whatever so that one can look at this reducer and know what the structure of this branch will be.

dmitryduev replied:
Ah, right! It will be relevant for the next PR where the filter page will be implemented, so I'll put it in there, if you don't mind.

acrellin replied:
Sure

acrellin commented on static/js/ducks/filter.js:
```diff
@@ -0,0 +1,104 @@
+const reducerV = (state = {}, action) => {
+  switch (action.type) {
+    case FETCH_FILTER_V_OK: {
+      return action.data;
```
Same comment here as above

acrellin commented on static/js/ducks/stream.js:
```diff
@@ -0,0 +1,44 @@
+const reducer = (state = {}, action) => {
+  switch (action.type) {
+    case FETCH_STREAM_OK: {
+      return action.data;
```
Same comment here as above in filter.js

dmitryduev replied:
It's a placeholder for a next PR that will introduce the Streams page, I would prefer to put it in there.

acrellin commented on static/js/ducks/streams.js:
```diff
@@ -0,0 +1,63 @@
+function reducer(state = null, action) {
+  switch (action.type) {
+    case FETCH_STREAMS_OK: {
+      return action.data;
```
Same comment here as above

dmitryduev replied:
It's a placeholder for a next PR that will introduce the Streams page, I would prefer to put it in there.

dmitryduev commented on static/js/components/Group.jsx:
```diff
@@ -1,18 +1,125 @@
+
   const [groupLoadError, setGroupLoadError] = useState("");
 
+  const [expanded1, setExpanded1] = React.useState("panel1");
```
Sure, sorry about that. Fixed.

dmitryduev commented on static/js/components/Group.jsx:
```diff
@@ -1,18 +1,125 @@
+    setAddStreamOpen(false);
+  };
+
+  const handleChange1 = (panel) => (event, isExpanded) => {
```
Fixed

dmitryduev commented on static/js/components/Group.jsx:
```diff
@@ -47,61 +200,378 @@ const Group = () => {
       }
     });
 
+    const group_stream_ids = group.streams.map((stream) => stream.id);
```
+1

dmitryduev commented on static/js/components/Group.jsx:
```diff
@@ -47,61 +200,378 @@ const Group = () => {
 
+    const group_stream_ids = group.streams.map((stream) => stream.id);
+
+    const isGoed = (sid) => streams.map((stream) => stream.id).includes(sid);
```
Oops, that's Dutch for "it is ok", I must have experimented with Controller validation rules and it must have unexpectedly worked from the first attempt :) Renamed into `isStreamIdInStreams`.

dmitryduev commented on static/js/components/Group.jsx:
```diff
@@ -47,61 +200,378 @@ const Group = () => {
+                      </IconButton>
+                    </ListItemSecondaryAction>
+                  )}
+                  {!isAdmin(user, group) && (
```
These are rendered for group admins so that they can delete individual users.

dmitryduev commented:
@acrellin thanks for the review! I addressed your comments and while doing so implemented some additional (permissioning) logic that was not there; will add a few more tests and should be ready for re-review.

stefanv commented on skyportal/tests/frontend/test_groups.py:
```diff
@@ -49,83 +49,141 @@ def test_add_new_group_explicit_self_admin(driver, super_admin_user, user):
+    )
+
+    # delete filter
+    delete_button = driver.wait_for_xpath(f'//a[contains(.,"{filter_name}")]')
```
click_xpath should do the trick here

stefanv replied:
Ah, you are finding multiple and only clicking the first; perhaps there is a way to do that with xpath, but also ok otherwise.

dmitryduev replied:
Right.

stefanv commented on static/js/components/Group.jsx:
```diff
@@ -1,18 +1,129 @@
+    paddingLeft: theme.spacing(2),
+  },
+  heading: {
+    fontSize: theme.typography.pxToRem(17),
```
Why not just specify in rem?

dmitryduev replied:
Sure, fixed

stefanv commented on static/js/components/Group.jsx:
```diff
@@ -1,18 +1,129 @@
+    flexDirection: "column",
+  },
+  button_add: {
+    maxWidth: "140px",
```
px -> rem

dmitryduev replied:
fixed

stefanv commented on static/js/components/Group.jsx:
```diff
@@ -1,18 +1,129 @@
+  },
+  paper: {
+    width: "100%",
+    padding: theme.spacing(1),
```
@acrellin Do you know if this is absolute or relative spacing?  I would guess absolute.

stefanv commented on static/js/components/Group.jsx:
```diff
@@ -1,18 +1,129 @@
+    "panel-streams"
+  );
+  const [dialogOpen, setDialogOpen] = React.useState(false);
+  const fullScreen = useMediaQuery(theme.breakpoints.down("md"));
```
Please read https://skyportal.io/docs/styling.html#responsive-design for caveats of down and why we prefer `up`.

dmitryduev replied:
Thanks for pointing me at it, fixed.

dmitryduev commented on skyportal/handlers/api/stream.py:
```diff
@@ -9,7 +9,7 @@
 
 class StreamHandler(BaseHandler):
-    @permissions(["System admin"])
+    @auth_or_token
```
I needed a less strict permissions check for some frontend stuff to work properly. I believe there is nothing criminal in knowing what streams are in the system. We can always put in a check on user's access to individual streams in the body of this method later (if requested).

dmitryduev commented on skyportal/tests/frontend/test_groups.py:
```diff
@@ -49,83 +49,141 @@ def test_add_new_group_explicit_self_admin(driver, super_admin_user, user):
 
 
+@pytest.mark.flaky(reruns=2)
+@pytest.mark.xfail(strict=False)
```
`driver.scroll_to_element_and_click(driver.wait_for_xpath(f'//button[contains(.,"Delete Group")]'))` does not always work as desired, and 2 re-runs is not always enough. Any know-how on what to do here @stefanv @acrellin?

dmitryduev commented on package.json:
```diff
@@ -26,6 +26,7 @@
     "mui-datatables": "3.3.1",
     "npm-check": "5.9.2",
     "papaparse": "5.2.0",
+    "prettier": "^2.1.0",
```
Done

stefanv commented on skyportal/tests/frontend/test_groups.py:
```diff
@@ -121,10 +127,51 @@ def test_delete_group_user(driver, super_admin_user, user, public_group):
+    driver.wait_for_xpath_to_be_clickable(
+        f'//li[contains(.,"{public_stream.id}")]', timeout=10
+    )
+    stream = driver.switch_to.active_element
```
@acrellin You resolved this, but I didn't see an explanation; do you have it?

dmitryduev replied:
@stefanv now that I think about it, it is quite bizarre actually... Selenium was complaining about stuff being on top of the  element in question, and this trick solved that issue, but I agree with you in that I don't understand how.

dmitryduev commented:
@acrellin could we merge this now?

acrellin commented:
@stefanv I didn't see that you had asked a question there, my apologies. I thought I was just resolving my own comment.

stefanv commented:
> @stefanv I didn't see that you had asked a question there, my apologies. I thought I was just resolving my own comment.

No worries, @acrellin, just thought I missed a discussion elsewhere.
